src/main/java/com/nutomic/syncthingandroid/fragments/DeviceFragment.java

```diff
@@ -0,0 +1,405 @@
+
+    private TextView mSyncthingVersionView;
+
+    private View mCompressionContainer;
+
+    private boolean mIsCreateMode;
+
+    private DialogInterface.OnClickListener mCompressionEntrySelectedListener = new DialogInterface.OnClickListener() {
+        @Override
+        public void onClick(DialogInterface dialog, int which) {
+            dialog.dismiss();
+
+            Compression compression = Compression.fromIndex(which);
+            // Don't pop the restart dialog unless the value is actually different.
+            if (compression != Compression.fromValue(getActivity(), mDevice.compression)) {
+                mDevice.compression = compression.getValue(getActivity());
+                mCompressionValueView.setText(compression.getTitle(getActivity()));
+                updateDevice();
+            }
+        }
+    };
+
+    private TextWatcher mIdTextWatcher = new TextWatcherAdapter() {
+        @Override
+        public void afterTextChanged(Editable s) {
+            mDevice.deviceID = s.toString();
+        }
+    };
+
+    private TextWatcher mNameTextWatcher = new TextWatcherAdapter() {
+        @Override
+        public void afterTextChanged(Editable s) {
+            mDevice.name = s.toString();
+        }
+    };
+
+    private TextWatcher mAddressesTextWatcher = new TextWatcherAdapter() {
+        @Override
+        public void afterTextChanged(Editable s) {
+            mDevice.addresses = persistableAddresses(s);
+        }
+    };
+
+    private CompoundButton.OnCheckedChangeListener mIntroducerCheckedChangeListener = new CompoundButton.OnCheckedChangeListener() {
+        @Override
+        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
+            mDevice.introducer = isChecked;
+            updateDevice();
+        }
+    };
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        SettingsActivity activity = (SettingsActivity) getActivity();
+        mIsCreateMode = activity.getIsCreate();
+        activity.registerOnServiceConnectedListener(this);
+        activity.setTitle(mIsCreateMode ? R.string.add_device : R.string.edit_device);
+        setHasOptionsMenu(true);
+
+        if (mIsCreateMode) {
+            if (savedInstanceState != null) {
+                mDevice = (RestApi.Device) savedInstanceState.getSerializable("device");
+            }
+            if (mDevice == null) {
+                initDevice();
+            }
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mSyncthingService != null) {
+            mSyncthingService.unregisterOnApiChangeListener(this);
+        }
+    }
+
+    @Override
+    public void onStop() {
+        super.onStop();
+
+        // We don't want to update every time a TextView's character changes,
+        // so we hold off until the view stops being visible to the user.
+        updateDevice();
```

Comment:
Oh wow. I hadn't tried to share but this sounds pretty weird. I'm pretty sure I have tried the introduced switch though and don't remember getting more than one dialogs. Could it be a combination of the two? You are right about the check, completely slipped my mind because the previous logic didn't need it. 

I was under the impression the API would handle the restart dialog display conditions and would not show it if the user has already dismissed another dialog previously. 
I will try to reproduce and fix these issues and try as many weird flows I can think of so we are extra safe. 

Sorry about this, major oversight.